### PR TITLE
@redwoodjs/history

### LINKED ISSUE #1892: RestoreAuthState no longer functioning correctly
When returning to the app from Auth0 the correct `targetUrl` from `appState` is applied to the address bar but the app doesn't actually route to it. Props at the home route have `code` and `state` and `pathname` from `useLocation` still shows `/`.

This appears to have started happening in the `0.22.0` release. Was working correctly in `0.21.0`. Still doesn't work correctly in the current release. I created a new project for testing to make sure it wasn't anything in our existing codebase causing the problem.

You can see from the following screenshots I just came back from Auth0 and I'm seeing the home page rather than the after auth page as defined in the routes.

Screen Shot 2021-03-01 at 5 01 00 PM

Screen Shot 2021-03-01 at 5 01 13 PM

Then if I refresh the browser I get the correct page as expected.

Screen Shot 2021-03-01 at 5 03 16 PM

### PR DESCRIPTION
**Pull request by Tobbe**
Don't pay too much attention to the code in this PR yet, it's a work in progress. But please do read through the text below

# The Problem

Some auth providers want to redirect the user when authorizing. They do this by a standard `window.history.replaceState` call (or `global.history.replaceState`). The address in the browser's address bar is updated, but the user isn't actually redirected.

# The Cause

A standard RW app might look something like this

https://github.com/redwoodjs/redwood-tutorial/blob/3b59bc12ca211688839b395b8d1364a4c111db26/web/src/App.js#L15-L23
```
const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <AuthProvider client={netlifyIdentity} type="netlify">
      <RedwoodApolloProvider>
        <Routes />
      </RedwoodApolloProvider>
    </AuthProvider>
  </FatalErrorBoundary>
)
```

Where `<Routes>` uses this code from the router

https://github.com/redwoodjs/redwood/blob/4abbe45da059029b317e1482466d962a998a8635/packages/router/src/router.tsx#L213-L225

A regular call to `navigate` looks like this

https://github.com/redwoodjs/redwood-tutorial/blob/2941c967adb6d7251ce3ae6e34edc3161e3fe89c/web/src/components/Post/Post.js#L36
```
navigate(routes.posts())
```

It will call this code where the address is updated (`history.pushState`) and all listeners are notified about the route change

https://github.com/redwoodjs/redwood/blob/4abbe45da059029b317e1482466d962a998a8635/packages/router/src/history.tsx#L13-L18

The listeners are located in `<LocationProvider>`. You can see that provider being used in the snippet from the router above. The provider's internal state is updated, and all its children are re-rendered. The router, being one of the children, sees the updated address and renders the new page ("posts" in this case).

## This, below, is the problem
The direct call to just `history.replaceState` by the auth client bypasses the `<LocationProvider>` update, and so no re-rendering of the router happens.

# The Solution

The solution might sound obvious. Just call the location listeners, like with a regular `navigate()` call.

We can't do this however, because the auth stuff is higher up in the hierarchy and so can't use the location context with its listeners.

## My proposal

I propose that we introduce a new package, "history", that's responsible for keeping track of history, history changes and history change notification. This would be placed above auth in the hierarchy so auth can update the history (by calling `history.replace`), and the LocationProvider can keep listening to changes to history, and re-render its children when it changes.

 Closes #1892 

---

Adding some more context from #1892 

This is what a typical App.js file looks like with auth0 added

```jsx
const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <AuthProvider client={auth0} type="auth0">
      <RedwoodApolloProvider>
        <Routes />
      </RedwoodApolloProvider>
    </AuthProvider>
  </FatalErrorBoundary>
)
```

When signing up you do something like this

```jsx
<button
  onClick={() =>
    signUp({
      appState: { targetUrl: '/after-auth' },
    })
  }
>
  Sign In
</button>
```

When `<AuthProvider>` from the `App` snippet above is mounted it calls `restoreAuthState`, which in turn does this

```js
restoreAuthState: async () => {
  if (
    global?.location?.search?.includes('code=') &&
    global?.location?.search?.includes('state=')
  ) {
    const { appState } = await client.handleRedirectCallback()
    history.replace(
      {},
      document.title,
      appState && appState.targetUrl
        ? appState.targetUrl
        : window.location.pathname
    )
  }
},
```

The problem is with the `history.replace` call. The router doesn't know about this (doesn't get any notifications). So the correct page isn't rendered. In fact, no rerender happens at all. So the user just stays on the current page. But the browser knows, so the URL is updated.

---

Here's a Github repo with a project that displays the issue https://github.com/tobbe/tobbe-redwood-history
Using the code in this PR the issue goes away. There's some info in the README in the repo on how to set it up.

### CODE REVIEW

**Comment by KrisCoulson** [comment]
@Tobbe love this!  I was thinking about picking up this smaller ticket but it might be worth lumping in with your new package. History back functionality: https://github.com/redwoodjs/redwood/issues/1013

**Comment by Tobbe** [comment]
@KrisCoulson I'm not going to have time to look at this for a while. Do you want to pick it up instead?

**Comment by sangheestyle** [comment]
Hi @Tobbe ,

Can we make a separate PR for the `jump` first? I use `global.history.replaceState()` when I need it but it might be good if rw router have one for it and the listeners can be used semlessly.

https://github.com/redwoodjs/redwood/blob/2ced364144b195a579c16b69ce8a96fe97ebf866/packages/router/src/history.tsx#L28

**Comment by Tobbe** [comment]
> Hi @Tobbe ,
> 
> Can we make a separate PR for the `jump` first? I use `global.history.replaceState()` when I need it but it might be good if rw router have one for it. So the listeners can be used seamlessly.

Hi @sangheestyle thanks for showing interest in this PR 🙂 I will spend some more time on this, and then discuss with the rest of the team about the best way forward.

**Comment by sangheestyle** [comment]
> > Hi @Tobbe ,
> > Can we make a separate PR for the `jump` first? I use `global.history.replaceState()` when I need it but it might be good if rw router have one for it. So the listeners can be used seamlessly.
> 
> Hi @sangheestyle thanks for showing interest in this PR 🙂 I will spend some more time on this, and then discuss with the rest of the team about the best way forward.

Awesome! Thanks for the reply!

**Comment by Tobbe** [comment]
@sangheestyle We decided to focus on getting this PR done as soon as possible instead of spending time on a separate PR for `jump()`. I don't want to make any promises, but we'll try to get this done this or next week.

**Comment by Tobbe** [comment]
Here's a RW project to reproduce the issue, and verify the fix: https://github.com/tobbe/tobbe-redwood-history

File: packages/auth/src/AuthProvider.tsx [hunk 1]
```diff
@@ -107,7 +107,8 @@ export class AuthProvider extends React.Component<
 
   async componentDidMount() {
     await this.rwClient.restoreAuthState?.()
```
Comment (Tobbe) [inline packages/auth/src/AuthProvider.tsx:109]: Probably shouldn't do this twice

**Review by jtoar — APPROVED** [review]

**Comment by thedavidprice** [comment]
@jtoar confirming there are no breaking changes from this PR, correct? Also, is there anything we should do for documentation? Assuming "no".

**Comment by jtoar** [comment]
@thedavidprice no breaking changes. But ideally this PR will be superseded by https://github.com/redwoodjs/redwood/pull/3055 before our next release.

We could add to the router documentation for the new `replace` option to navigate and  the `back` function.

**Pull request merged**